eido/conversion.py

```diff
@@ -51,17 +52,43 @@ def run_filter(prj, filter_name, plugin_kwargs=None):
     :param dict plugin_kwargs: kwargs to pass to the plugin function
     :raise EidoFilterError: if the requested filter is not defined
     """
+    # convert to empty dictionary if no plugin_kwargs are passed
+    plugin_kwargs = plugin_kwargs or dict()
+
     installed_plugins = pep_conversion_plugins()
     installed_plugin_names = list(installed_plugins.keys())
+    path = plugin_kwargs.get("path")
+
     if filter_name not in installed_plugin_names:
         raise EidoFilterError(
             f"Requested filter ({filter_name}) not found. "
             f"Available: {', '.join(installed_plugin_names)}"
         )
     _LOGGER.info(f"Running plugin {filter_name}")
     func = installed_plugins[filter_name]
-    plugin_kwargs = plugin_kwargs or dict()
-    func(prj, **plugin_kwargs)
+
+    conv_result = func(prj, **plugin_kwargs)
+
+    if path is not None:
+        # make out dir if it doesn't
+        # already exist
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+        # loop through and write necessary
+        # files to disk
+        for out_file in conv_result:
```

Comment:
Here you are getting the output file from the filter, and then treating `path` as a single folder.

I thought what we had decided to do was accept a dict of `paths`, and then use those, by linking to keys of the return types.

So in other words, a filter with multiple outputs would specify in its readme, "I provide 2 output channels, named `cfg` and `samples`. The output object would look like this

```
return  {
  "cfg": cfg_string,
  "samples" : sampels_csv_string
}
```

The input would have to be something like this:

```
paths = {
  "cfg": "/path/to/config.yaml",
  "samples" : "/path/to/samples.csv"
}
```

So, users are instructed to use the filter, they have to provide 2 paths, keyed with `cfg` and `samples`, in order to use this filter. Those keyword are specified by the filter and explained in the filter documentation.

The `run_filter` function matches `cfg` from the `paths` object, provided by the user in `paths` in the `kwargs`, to the `cfg` key of the return value of the filter, and prints `cfg_string` to "/path/to/config.yaml" -- and the same thing for `samples`.

Reply:
Also just to give some motivation... the advantage of the way I propose over the way it's currently written is that the person who calls the filter has more control over the output. As it is, the person calling the filter cannot control the filenames of the output, and also is restricted to putting all the outputs into the same folder. The revision I propose removes responsibility of naming the files from the filter function, and puts that responsibility on the person calling the filter. That's the person who's more likely to know where the files should go and what they should be named.

It also cleans up the output of the filter functions, just semantically. The result may not even be destined for a file at all, as in pephub -- in that case, naming the output dict with filenames as keys makes no sense. These are really just strings we're returning -- strings with keys that describe what they are. The user can put them in a file... or not.

Reply:
#39